lib/SILOptimizer/Analysis/ARCAnalysis.cpp

```diff
@@ -1154,3 +1156,132 @@ SILInstruction *swift::findReleaseToMatchUnsafeGuaranteedValue(
   }
   return nullptr;
 }
+
+void EpilogueARCContext::initializeDataflow() {
+  // Find the exit blocks.
+  auto Throw = F->findThrowBB();
+  auto Return = F->findReturnBB();
+  if (Throw != F->end())
+    ExitBlocks.insert(&*Throw);
+  if (Return != F->end())
+    ExitBlocks.insert(&*Return);
+  // Return and throw blocks BBSetOut is initialized to true.
+  for (auto &B : *F) {
+    EpilogueARCBlockStates[&B] =
+              new (BPA.Allocate()) EpilogueARCBlockState();
+  }
+}
+
+void EpilogueARCContext::convergeDataflow() {
+  // Keep iterating until Changed is false.
+  bool Changed = false;
+  do {
+    Changed = false;
+    // Iterate until the data flow converges.
+    for (SILBasicBlock *B : PO->getPostOrder()) {
+      auto BS = EpilogueARCBlockStates[B];
+      // Merge in all the successors.
+      if (!B->succ_empty()) {
+        auto Iter = B->succ_begin();
+        BS->BBSetOut = EpilogueARCBlockStates[*Iter]->BBSetIn;
+        Iter = std::next(Iter);
+        for (auto E = B->succ_end(); Iter != E; ++Iter) {
+	   BS->BBSetOut &= EpilogueARCBlockStates[*Iter]->BBSetIn;
+        }
+      } else if (isExitBlock(B)) {
+        // We set the BBSetOut for exit blocks.
+        BS->BBSetOut = true;
+      }
+
+      // If an epilogue ARC instruction or blocking operating has been identified
+      // then there is no point visiting every instruction in this block.
+      if (!BS->BBSetOut) {
+        Changed |= (BS->BBSetIn != BS->BBSetOut);
+        BS->BBSetIn = BS->BBSetOut;
+        continue;
+      }
+
+      // Iterate over all instructions in the basic block and find the
+      // interested ARC instruction in the block.
+      for (auto I = B->rbegin(), E = B->rend(); I != E; ++I) {
+        // This is a transition from 1 to 0 due to an interested instruction.
+        if (isInterestedInstruction(&*I)) {
+          BS->BBSetOut = false;
+          break;
+        }
+        // This is a transition from 1 to 0 due to a blocking instruction.
+        if (mayBlockEpilogueARC(&*I, RCFI->getRCIdentityRoot(Arg))) {
+          BS->BBSetOut = false;
+          break;
+        }
+      }
+
+      // Update BBSetIn.
+      Changed |= (BS->BBSetIn != BS->BBSetOut);
+      BS->BBSetIn = BS->BBSetOut;
+    }
+  } while(Changed);
+}
+
+bool EpilogueARCContext::computeEpilogueARC() {
+  // At this point the data flow should have converged. Find the epilogue
+  // releases.
+  for (SILBasicBlock *B : PO->getPostOrder()) {
+    auto BS = EpilogueARCBlockStates[B];
+    // Merge in all the successors.
+    if (!B->succ_empty()) {
+      // Make sure we've either found no ARC instructions in all the successors
+      // or we've found ARC instructions in all successors.
+      auto Iter = B->succ_begin();
+      auto Base = EpilogueARCBlockStates[*Iter]->BBSetIn;
+      Iter = std::next(Iter);
+      for (auto E = B->succ_end(); Iter != E; ++Iter) {
+        if (EpilogueARCBlockStates[*Iter]->BBSetIn != Base)
+          return false;
```

Comment:
This return false is really difficult to understand. Maybe you can clarify this (in addition to your comment above), e.g. add an example.